cudax/include/cuda/experimental/__memory_resource/cuda_async_memory_resource.h

```diff
@@ -0,0 +1,287 @@
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+// cudaMallocAsync was introduced in CTK 11.2
+#if !defined(_CCCL_COMPILER_MSVC_2017) && !defined(_CCCL_CUDACC_BELOW_11_2)
+
+#  if !defined(_CCCL_CUDA_COMPILER_NVCC) && !defined(_CCCL_CUDA_COMPILER_NVHPC)
+#    include <cuda_runtime.h>
+#    include <cuda_runtime_api.h>
+#  endif // !_CCCL_CUDA_COMPILER_NVCC && !_CCCL_CUDA_COMPILER_NVHPC
+
+#  include <cuda/__memory_resource/get_property.h>
+#  include <cuda/__memory_resource/properties.h>
+#  include <cuda/__memory_resource/resource_ref.h>
+#  include <cuda/std/__concepts/__concept_macros.h>
+#  include <cuda/std/__cuda/api_wrapper.h>
+#  include <cuda/std/__new_>
+#  include <cuda/std/cstddef>
+#  include <cuda/stream_ref>
+
+#  include <cuda/experimental/__memory_resource/cuda_memory_pool.h>
+
+#  if _CCCL_STD_VER >= 2014
+
+namespace cuda::experimental::mr
+{
+
+//! @rst
+//! .. _cudax-memory-resource-async:
+//!
+//! Stream ordered memory resource
+//! ------------------------------
+//!
+//! ``cuda_async_memory_resource`` uses `cudaMallocFromPoolAsync / cudaFreeAsync
+//! <https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__MEMORY__POOLS.html>`__ for allocation/deallocation. A
+//! ``cuda_async_memory_resource`` is a thin wrapper around a \c cudaMemPool_t.
+//!
+//! .. warning::
+//!
+//!    ``cuda_async_memory_resource`` does not own the pool and it is the responsibility of the user to ensure that the
+//!    lifetime of the pool exceeds the lifetime of the ``cuda_async_memory_resource``
+//!
+//! @endrst
+class cuda_async_memory_resource
+{
+private:
+  ::cudaMemPool_t __pool_;
+
+  //! @brief Checks whether the passed in alignment is valid
+  //! @param __alignment the alignment to check
+  //! @returns true if \p __alignment is valid
+  _CCCL_NODISCARD static constexpr bool __is_valid_alignment(const size_t __alignment) noexcept
+  {
+    return __alignment <= _CUDA_VMR::default_cuda_malloc_alignment
+        && (_CUDA_VMR::default_cuda_malloc_alignment % __alignment == 0);
+  }
+
+  //! @brief  Returns the default ``cudaMemPool_t`` from the current device
+  //! @throws cuda_error if retrieving the default ``cudaMemPool_t`` fails
+  //! @returns The default memory pool of the current device
+  _CCCL_NODISCARD static ::cudaMemPool_t __get_default_mem_pool()
+  {
+    const auto __device_id = _CUDA_VMR::__get_current_cuda_device();
+    _LIBCUDACXX_ASSERT(_CUDA_VMR::__device_supports_pools(__device_id), "cudaMallocAsync not supported");
+
+    ::cudaMemPool_t __pool;
+    _CCCL_TRY_CUDA_API(
+      ::cudaDeviceGetDefaultMemPool, "Failed to call cudaDeviceGetDefaultMemPool", &__pool, __device_id);
+    return __pool;
+  }
+
+public:
+  //! @brief Default constructs the cuda_async_memory_resource using the default \c cudaMemPool_t of the current device
+  //! @throws cuda_error if retrieving the default \c cudaMemPool_t fails
+  cuda_async_memory_resource()
```

Comment:
Should this also have a constructor taking `device_ref` to explicitly set the device from which to obtain the default pool? As I understand it, the memory pool overloads always create a new pool, so having a way to just use a specific device's default pool without having to switch to that device seems like a good feature.

Reply:
that is a great suggestion thanks